FV3/atmos_cubed_sphere/tools/coarse_grained_restart_files.F90

```diff
@@ -0,0 +1,366 @@
+    type(domain2d), intent(in) :: coarse_domain
+    type(coarse_restart_type), intent(inout) :: restart
+
+    character(len=64) :: filename
+    integer :: id_restart
+
+    filename = 'fv_land_coarse.res.nc'
+
+    id_restart = register_restart_field(restart%fv_land_coarse, &
+         filename, 'oro', restart%oro, domain=coarse_domain, &
+         tile_count=tile_count)
+  end subroutine register_fv_land_coarse
+
+  subroutine coarse_grain_restart_data(Atm)
+    type(fv_atmos_type), intent(inout) :: Atm
+
+    if (trim(Atm%coarse_graining%strategy) .eq. MODEL_LEVEL) then
+       call coarse_grain_restart_data_on_model_levels(Atm)
+    endif
+  end subroutine coarse_grain_restart_data
+  
+  subroutine coarse_grain_restart_data_on_model_levels(Atm)
+    type(fv_atmos_type), intent(inout) :: Atm
+
+    real, allocatable :: mass(:,:,:)
+
+    allocate(mass(is:ie,js:je,1:npz))
+    call compute_mass_weights(Atm%gridstruct%area(is:ie,js:je), Atm%delp(is:ie,js:je,1:npz), mass)
+
+    call coarse_grain_fv_core_restart_data_on_model_levels(Atm, mass)
+    call coarse_grain_fv_tracer_restart_data_on_model_levels(Atm, mass)
+    call coarse_grain_fv_srf_wnd_restart_data_on_model_levels(Atm)
+    if (Atm%flagstruct%fv_land) then
+       call coarse_grain_mg_drag_restart_data_on_model_levels(Atm)
+       call coarse_grain_fv_land_restart_data_on_model_levels(Atm)
+    endif
+  end subroutine coarse_grain_restart_data_on_model_levels
+
+  subroutine coarse_grain_fv_core_restart_data_on_model_levels(Atm, mass)
+    type(fv_atmos_type), intent(inout) :: Atm
+    real, intent(in) :: mass(is:ie,js:je,1:npz)
+
+    call weighted_block_edge_average_x(Atm%gridstruct%dx(is:ie,js:je+1), &
+         Atm%u(is:ie,js:je+1,1:npz), Atm%coarse_graining%restart%u)
+    call weighted_block_edge_average_y(Atm%gridstruct%dy(is:ie+1,js:je), &
+         Atm%v(is:ie+1,js:je,1:npz), Atm%coarse_graining%restart%v)
+
+    if (.not. Atm%flagstruct%hydrostatic) then
+       call weighted_block_average(mass(is:ie,js:je,1:npz), &
+            Atm%w(is:ie,js:je,1:npz), Atm%coarse_graining%restart%w)
+       call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+            Atm%delz(is:ie,js:je,1:npz), Atm%coarse_graining%restart%delz)
+       if (Atm%flagstruct%hybrid_z) then
+          call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+            Atm%ze0(is:ie,js:je,1:npz), Atm%coarse_graining%restart%ze0)
+       endif
+    endif
+
+    call weighted_block_average(mass(is:ie,js:je,1:npz), &
+         Atm%pt(is:ie,js:je,1:npz), Atm%coarse_graining%restart%pt)
+    call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+         Atm%delp(is:ie,js:je,1:npz), Atm%coarse_graining%restart%delp)
+    call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
+         Atm%phis(is:ie,js:je), Atm%coarse_graining%restart%phis)
+
+    if (Atm%flagstruct%agrid_vel_rst) then
+       call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), &
```

Comment:
I wonder whether we want to weight by mass here? It would be inconsistent with the D-grid wind coarsening, but presumably closer to momentum-conserving. Is the reason we just weight by edge length for the D-grid wind coarsening because we don't have easy access to the `delp` on the D-grid?

Reply:
I recall the edge-length-weighted average was recommended by Lucas, presumably due to an analogy with how he handled things in the nesting code.  I went with an area-weighted average somewhat by analogy here.  

I do think this could be worth checking again with Lucas about.  When I get a chance I'll send him an email.

Reply:
As discussed during standup, this choice isn't meaningful when we switch to pressure-level coarse-graining (what we will ultimately use), so we probably don't need to think too hard about it.  I'll change things to mass-weighted averaging here, since it does make difference when coarsening on model levels, and that seems to be the consensus (though we'll still see what Lucas says).

Reply:
Yes, that's a good point about it not making a difference when doing pressure-level. I agree with changing to mass-weighting for coarsening on model levels.